core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -199,6 +201,8 @@ protected void validate(TableMetadata currentMetadata, Snapshot snapshot) {}
    */
   protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate, Snapshot snapshot);
 
+  protected abstract Long startingSnapshotId();
```

Comment:
I think this has to be non null? For the cases which it is currently null, it's just because the value is missing. But we always know what is the starting snapshot ID of the commit.

Reply:
are you suggesting the `starting snapshot ID` will be the base on which we are applying the snapshot ?